src/services/fireblocks.service.js

```diff
@@ -0,0 +1,369 @@
+const jwt = require('jsonwebtoken');
+
+const { getConfig } = require('../settings');
+
+const FIREBLOCKS_VERSION = 'v1';
+
+const TX_ALREADY_EXISTS_ERROR_CODE = 1438;
+
+/**
+ * Encode Fireblocks signature in DER
+ * @param {object} sig
+ * @returns {Buffer}
+ */
+function fireblocksSignatureToDER(sig) {
+  const r = Buffer.from(sig.r, 'hex');
+  const s = Buffer.from(sig.s, 'hex');
+
+  const rneg = !!(r[0] & 0x80);
+  const sneg = !!(s[0] & 0x80);
+
+  const rbuf = rneg ? Buffer.concat([Buffer.from([0x00]), r]) : r;
+  const sbuf = sneg ? Buffer.concat([Buffer.from([0x00]), s]) : s;
+
+  const rlen = rbuf.length;
+  const slen = sbuf.length;
+  const length = 2 + rlen + 2 + slen;
+
+  // 0x30 is a sequence header followed by the length in bytes of the sequence (with headers)
+  // 0x02 is an integer header followed by the length in bytes of the integer then the integer
+  return Buffer.concat([
+    Buffer.from([0x30, length, 0x02, rlen]),
+    rbuf,
+    Buffer.from([0x02, slen]),
+    sbuf,
+  ]);
+}
+
+/**
+ * Get signatures of the transaction for the wallet in the storage
+ * @param {hathorLib.Transaction} tx
+ * @param {hathorLib.Storage} storage
+ * @param {FireblocksClient} client
+ */
+async function getTxSignatures(tx, storage, client) {
+  const dataToSignHash = tx.getDataToSignHash();
+  const inputSignatures = [];
+  let ncCallerSignature = null;
+  let ncCallerIndex = null;
+  const signingIndices = [];
+
+  for await (const { tx: spentTx, input, index: inputIndex } of storage.getSpentTxs(tx.inputs)) {
+    if (input.data) {
+      // This input is already signed
+      continue;
+    }
+
+    const spentOut = spentTx.outputs[input.index];
+
+    if (!spentOut.decoded.address) {
+      // This output is not owned
+      continue;
+    }
+    const addressInfo = await storage.getAddressInfo(spentOut.decoded.address);
+    if (!addressInfo) {
+      // Not a wallet address
+      continue;
+    }
+
+    const pubkeyHex = await storage.getAddressPubkey(addressInfo.bip32AddressIndex);
+    const pubkey = Buffer.from(pubkeyHex, 'hex');
+
+    signingIndices.push({
+      inputIndex,
+      addressIndex: addressInfo.bip32AddressIndex,
+      pubkey,
+    });
+  }
+
+  /* istanbul ignore next */
+  if (tx.version === hathorLib.constants.NANO_CONTRACTS_VERSION) {
+    const { pubkey } = tx;
+    const address = hathorLib.addressUtils.getAddressFromPubkey(pubkey.toString('hex'), storage.config.getNetwork());
+    const addressInfo = await storage.getAddressInfo(address.base58);
+    if (addressInfo) {
+      ncCallerIndex = addressInfo.bip32AddressIndex;
+    }
+  }
+
+  // Now we sign
+  const indices = signingIndices.map(data => data.addressIndex);
+  if (ncCallerIndex) {
+    indices.push(ncCallerIndex);
+  }
+
+  // create signature data object from the response
+  const fbTxInfo = await client.sendRawTransaction(dataToSignHash, indices);
+  if (!fbTxInfo.signedMessages) {
+    throw new Error('Fireblocks did not return signatures');
```

Comment:
question(non-blocking): Would it be a good idea to log any information from this `fbTxInfo` to help with debugging? Maybe the Fireblocks transaction id?

Reply:
Done